### PR TITLE
KOGITO-4265: [DMN Designer] Decision Services - The parameters order in the properties panel is not correct

### PR DESCRIPTION
**Pull request by danielzhe**
**JIRA**: [KOGITO-4265](https://issues.redhat.com/browse/KOGITO-4265)

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

### CODE REVIEW

**Comment by danielzhe** [comment]
jenkins do fdb

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/decisionservice/parameters/DecisionServiceParametersListWidget.java [hunk 1]
```diff
@@ -151,10 +151,24 @@ void refresh() {
...
+        getValue().getDecisionService().getInputData().forEach(ref -> {
+            final String href = ref.getHref().replace("#", "");
+            final Optional<InputData> currentInput = inputs.stream()
+                    .filter(input -> Objects.equals(input.getId().getValue(), href))
+                    .findFirst();
+            currentInput.ifPresent(inputData -> sorted.add(inputData));
+        });
+
+        return sorted;
+    }
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/decisionservice/parameters/DecisionServiceParametersListWidget.java:170]: I think we should javadoc what sorting is done, I had in mind alphabetical but that is not the case.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/decisionservice/parameters/DecisionServiceParametersListWidgetTest.java [hunk 2]
```diff
@@ -191,10 +195,64 @@ public void testRefresh() {
...
+                                                             createInputDataWithId(id1));
+
+        doReturn(value).when(widget).getValue();
+
+        when(value.getDecisionService()).thenReturn(decisionService);
+        when(decisionService.getInputData()).thenReturn(list);
+
+        final List<InputData> sorted = widget.getSortedInputs(unsortedInputs);
+
+        checkIfInputDataListIsInOrder(sorted, id1, id2, id3);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/widgets/decisionservice/parameters/DecisionServiceParametersListWidgetTest.java:227]: Please do not introduce this private method. We have already on class path assertj library with implemented assertions for  checking collection elements are in order

```java
Assertions.assertThat(sorted)
                .extracting(inputData -> inputData.getId().getValue())
                .containsExactly(id1, id2, id3);
```

**Review by jomarko — CHANGES_REQUESTED** [review]
Running business central war, I needed to reopen the dmn model to see actual input nodes.

Saved but not all inputs shown
Screenshot from 2021-01-27 23-03-54

Reopened, all inputs shown
Screenshot from 2021-01-27 23-04-13


I can confirm order in UI is same as in stored xml, however previously there was no need to reopn file to see all inputs.

**Comment by danielzhe** [comment]
I'm unable to reproduce the issue you reported. For me looks like it is working in BC. I change the inputs and when I select Decision Service again the widget is updated. Can you provide me a step-by-step or a video to reproduce? Thank you, @jomarko

**Comment by jomarko** [comment]
@danielzhe yes, [here](https://issues.redhat.com/browse/KOGITO-4265?focusedCommentId=15715660&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15715660) is what I see once download war form downstream build. 

If you will be not able to reproduce, could you please rebase your branch with master - it is already 7.50.0-SNAPSHOT and relaunch the downstream job please? Thank you.

**Comment by danielzhe** [comment]
@jomarko Thank you! Changes applied.

**Comment by danielzhe** [comment]
jenkins do fdb

**Review by jomarko — COMMENTED** [review]
@danielzhe thank you for the updates. I can confirm I do not see the issue reported in [comment above](https://github.com/kiegroup/kie-wb-common/pull/3549#issuecomment-769881289).

However there is some strange corner case related to first close and reopen.
If I create a diagram:
1. Add a decision service node
2. Add a decision node into the decision service node
3. Add an input data node, rename it to 'Z'
4. Add an input data node, rename it to 'Y'
5. Add an input data node, rename it to 'X'
6. Connect 'Z' with the decision node
7. Connect 'Y' with the decision node
8. Connect 'X' with the decision node
9. Save the model
10. Check the properties panel, order of the decision service inputs will be, 'Z', 'Y', 'X'
11. Close and reopen the model
12. Check the properties panel, order of the decision service inputs will be, 'X', 'Y', 'Z'
13. Afterwards the order seem to be always correct

I attached screenshots and the source of the model also into the jira, see 'a.dmn' model.

**Comment by jomarko** [comment]
@karreiro as a reporter, is there anything you have in mind could be related to fact the behavior seems be not correct until first model reopen is done? See [my comment](https://github.com/kiegroup/kie-wb-common/pull/3549#pullrequestreview-585961731).

**Comment by karreiro** [comment]
Thank you for reporting that, @jomarko.

@danielzhe @jomarko, the order of the parameters we're persisting in the decision service XML represents the order that users need to know to call the decision service in a literal expression correctly.

So, the scenario that @jomarko spotted is not what we expect. The UI component's expected behavior is always to show 'X', 'Y', 'Z' in steps 10 and 12 [1], thus users will know how to invoke the decision service correctly.

[1] Considering the decision service is persisted like this:

```xml
<dmn:decisionService>
    ...
    <dmn:inputData href="X"/>
    <dmn:inputData href="Y"/>
    <dmn:inputData href="Z"/>
</dmn:decisionService>
```

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by danielzhe** [comment]
@jomarko @karreiro Changes applied. The marshalling save new inputs in alphabetical order, so now I'm reflecting what marshalling does.

**Review by jomarko — COMMENTED** [review]
@danielzhe not sure why, on VS Code extension are the 'Input Data' of the decision service still not sorted not sorted.

In business central works fine immediately after opening the file.

I created also a reproducer test https://github.com/kiegroup/kie-wb-common/pull/3561

Screenshot from 2021-02-12 15-59-34

**Comment by danielzhe** [comment]
I was unable to reproduce the issue you reported, @jomarko. Looks like there's a change in `kogito-tooling` and the way we target specific versions of `kie-wb-common`.: https://github.com/kiegroup/kogito-tooling/#develop
At first I was trying to only build and I got the same issue you reported, but then I realized that possible I was not using the right `kie-wb-common` version so I did what is described on those steps (targeting my local build of this PR) and it worked. Maybe that is the issue.

**Review by karreiro — APPROVED** [review]
Thank you, @danielzhe!

@jomarko I just built a plugin with this PR with https://github.com/kiegroup/kogito-tooling/pull/417 (then we're able to save the DMN model), and it couldn't reproduce the issue locally, but I'm not sure if I'm missing something. Here's the [plugin](https://www.dropbox.com/s/c5gtn49wcrnx9wu/KOGITO-4265.vsix?dl=0).

**Review by jomarko — APPROVED** [review]
@danielzhe @karreiro 

Sorry, this took longer time to me. For me was confusing 'sorting' we use. Tha is why my test https://github.com/kiegroup/kie-wb-common/pull/3561 was failing. Now it is updated.

I thought all 'Decision Service' inputs are sorted when marshalling is executed. However that is not true we sort just those added after opening a diagram.

However I don't assume it as an issue. The order in UI  and stored XML/DMN seems to be consistent.

@danielzhe thank you. Approving.


Yes I will ask you for review of https://github.com/kiegroup/kie-wb-common/pull/3561 once this PR is merged  and I will rebase it to contain single commit.

**Pull request merged**